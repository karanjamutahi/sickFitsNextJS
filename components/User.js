import React, { Component } from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';


const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        me{
            id
            email
            firstname
            orders{
                id
            }
            cart {
                id
                quantity
                item {
                    id
                    price
                    image
                    title
                    description
                }
            }
        }
    }
`;

const User = (props) => (
    <Query query={CURRENT_USER_QUERY} {...props}>
        { 
            (payload) => {
                return props.children(payload);
            } 
        }
    </Query>
);

//Render Prop whose only child can be a function 
User.propTypes = {
    children: PropTypes.func.isRequired,
};
 
export default User;
export { CURRENT_USER_QUERY };