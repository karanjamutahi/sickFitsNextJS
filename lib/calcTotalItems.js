function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

export default totalItems;