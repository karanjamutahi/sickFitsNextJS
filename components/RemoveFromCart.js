import React from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';
import swal from 'sweetalert';

const REMOVE_FROM_CART_MUTATION = gql`
    mutation REMOVE_FROM_CART($id: ID!) {
        removeFromCart(id: $id) {
            id
            user {
                id
            }
        }
    }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`;

class RemoveFromCart extends React.Component {
    static proptypes = {
        id: PropTypes.string.isRequired,
    };
    
    update = (cache, payload) => {  //Gets called as soon as response comes back from server after Mutation has run
        //1. Read cache
        const data = cache.readQuery({ query: CURRENT_USER_QUERY});
        
        //2. Remove item from cache
        const cartItemID = payload.data.removeFromCart.id;
        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemID);
        
        //3. Write cache copy back
        cache.writeQuery({ query: CURRENT_USER_QUERY, data});
    }

    render() {
        return (
            <Mutation mutation={REMOVE_FROM_CART_MUTATION} variables={{id: this.props.id}} update={this.update} optimisticResponse={
                {
                __typename: 'Mutation',
                removeFromCart: {
                    id: this.props.id,
                    __typename: 'CartItem',
                    user: {
                        __typename: 'User',
                        id: 'No IDea'
                    }
                },
            }}>
                {(removeFromCart, { loading, error}) => {
                    return(
                        <BigButton disabled={loading} onClick={() => {
                            removeFromCart().catch(err => swal("Error", err.message, "error", {timer: 3000}));
                        }} title={"Delete Item"}>&times;</BigButton>
                    )}}
            </Mutation>
        );
    }
}

export default RemoveFromCart;