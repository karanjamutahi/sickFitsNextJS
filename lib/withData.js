import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : devEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cartopen value from cache
             const { cartOpen } = cache.readQuery({
               query: LOCAL_STATE_QUERY,
             });
             //Write the cart State to the opposite
             const data = {
               data: {
                 cartOpen: ! cartOpen
               }
             };
             cache.writeData(data);
             return data; 
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
