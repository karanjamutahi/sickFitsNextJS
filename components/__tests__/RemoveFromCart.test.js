import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Cart, { LOCAL_STATE_QUERY } from '../Cart';
import { CURRENT_USER_QUERY } from '../User';
import { fakeUser, fakeCartItem, fakeItem } from '../../lib/testUtils';
import { ApolloConsumer } from 'react-apollo';
import RemoveFromCart,{ REMOVE_FROM_CART_MUTATION } from '../RemoveFromCart';
import { ApolloClient } from 'apollo-boost';


const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: {
             ...fakeUser(),
            cart: [ fakeItem({id: 'abc123'})]
            }}}
    },
    {
        request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: 'abc123'}},
        result: { data: { removeFromCart: { 
            __typename: 'CartItem',
            id: 'abc123'
        }}}
    }
];

describe("RemoveFromCart Component", () => {
    it("Renders and matches Snapshot", () => {
        const wrapper = mount(
            <MockedProvider>
                <RemoveFromCart id="abc123"></RemoveFromCart>
            </MockedProvider>
        );
        //console.log(wrapper.debug());
        expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    });

    xit("Removes item from cart", async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {(client) => {
                        apolloClient = client;
                        return(<RemoveFromCart id="abc123"></RemoveFromCart>);
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );
        //console.log(apolloClient);
        const res = await apolloClient.query({query: CURRENT_USER_QUERY});
        console.log(res);
        expect(res.data.me.cart).toHaveLength(1);
        wrapper.find('button').simulate('click');
        await wait();
        const res2 = await apolloClient.query({query: CURRENT_USER_QUERY});
        console.log(res2); 
        expect(res2.data.me.cart).toHaveLength(0);
    });
});