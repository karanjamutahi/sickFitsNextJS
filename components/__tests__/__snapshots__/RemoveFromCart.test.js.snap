// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveFromCart Component Renders and matches Snapshot 1`] = `
<button
  className="RemoveFromCart__BigButton-emvtd6-0 gPWNSn"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Ã—
</button>
`;
